### PR TITLE
resolved #91: automatically generate gitea version

### LINKED ISSUE #91: Automatically generated version on build
I think we can learn Golang's version management. 

On master, all the version numbers will be gave on build and the version number is the commit hash. Since the hash is too long, we can use the first 7 characters.  i.e. tip+2b445c7

On release branch, all the version number will be gave on build and the version number is branch name plus git hash. i.e. 1.0+2b445c7

For a tag version i.e. 1.0.0, the version number could be static and we can give it a special number not a git hash.

So that, when a user report an issue, we can know the question is from which version quickly.

### PR DESCRIPTION
**Pull request by lunny**
And I create the make.bash for linux and mac, we need another make.bat for windows. Of course, this is a beginning of make bash. I think a bash file to make is simple and better than Makefile.

### CODE REVIEW

**Comment by tboerger** [comment]
I personally prefer a makefile, most of the bash content is done there.

File: make.bash [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env bash
+
+# Copyright 2016 The Gitea Authors. All rights reserved.
+# Use of this source code is governed by a MIT-style
+# license that can be found in the LICENSE file.
+
+version="unknow"
+
+if [ -f VERSION ]; then
+	cat /etc/passwd | read version
```
Comment (tboerger) [inline make.bash]: passwd?

File: make.bash [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    echo "no VERSION found and not a git project"
+    exit 1
+fi
+
+version=$(git rev-parse --abbrev-ref HEAD)
+tag=$(git describe --tag --always)
+
+if [ "$version" != "HEAD" ]; then
+    if [ "$version" == "master" ]; then
+        go build -ldflags "-X main.Version=tip+${tag}"
```
Comment (tboerger) [inline make.bash:26]: I'm not a fan of tip as a "version"

**Comment by makhov** [comment]
What's wrong with Makefile? Why we need make.bash?

**Comment by lunny** [comment]
You have to install extra make system. But for a bash, it has no more dependency.

**Comment by tboerger** [comment]
To be serious, I think nearly all people working with Go got make installed. I personally don't really want to maintain 2 files for a single use case.

**Comment by tboerger** [comment]
And if we change the version assignment I would also drop https://github.com/go-gitea/gitea/blob/master/modules/setting/setting.go#L51-L52 and the places where it gets used.

```
# grep -R BuildGitHash *
Makefile:15:LDFLAGS += -X "github.com/go-gitea/gitea/modules/setting.BuildGitHash=$(SHA)"
modules/setting/setting.go:52:  BuildGitHash string
modules/setting/setting.go:620:     log.Info("Build Git Hash: %s", BuildGitHash)
```

```
# grep -R BuildTime *
Makefile:14:LDFLAGS += -X "github.com/go-gitea/gitea/modules/setting.BuildTime=$(DATE)"
modules/setting/setting.go:51:  BuildTime    string
modules/setting/setting.go:618: if len(BuildTime) > 0 {
modules/setting/setting.go:619:     log.Info("Build Time: %s", BuildTime)
```

**Comment by tboerger** [comment]
And this is my suggestion for the Makefile to get it working based on this proposal, I would always keep a valid version number, `tip` is not a version number and with a plain `0.0.0` we can still always see that this have been built from master:

``` patch
--- Makefile.current    2016-11-07 09:25:49.668272844 +0100
+++ Makefile    2016-11-07 09:33:32.880272024 +0100
@@ -4,31 +4,30 @@
 EXECUTABLE := gitea
 IMPORT := github.com/go-gitea/gitea

-SHA := $(shell git rev-parse --short HEAD)
-DATE := $(shell date -u '+%Y-%m-%d %I:%M:%S %Z')
-
 BINDATA := $(shell find conf | sed 's/ /\\ /g')
 STYLESHEETS := $(wildcard public/less/index.less public/less/_*.less)
 JAVASCRIPTS :=

-LDFLAGS += -X "github.com/go-gitea/gitea/modules/setting.BuildTime=$(DATE)"
-LDFLAGS += -X "github.com/go-gitea/gitea/modules/setting.BuildGitHash=$(SHA)"
+REF ?= $(shell git rev-parse --abbrev-ref HEAD)
+SHA ?= $(shell git describe --tag --always)

 TARGETS ?= linux/*,darwin/*,windows/*
 PACKAGES ?= $(shell go list ./... | grep -v /vendor/)

 TAGS ?=

-ifneq ($(TRAVIS_TAG),)
-   VERSION ?= $(TRAVIS_TAG)
+ifneq ($(REF),HEAD)
+   VERSION ?= 0.0.0+$(SHA)
 else
-   ifneq ($(TRAVIS_BRANCH),)
-       VERSION ?= $(TRAVIS_BRANCH)
+   ifneq ($(REF),master)
+       VERSION ?= 0.0.0+$(SHA)
    else
-       VERSION ?= master
+       VERSION ?= $(REF)+$(SHA)
    endif
 endif

+LDFLAGS += -X "main.Version=$(VERSION)"
+
 .PHONY: all
 all: clean test build
```

**Comment by strk** [comment]
What's wrong with just using SHA ? Why do we need REF too ? Do you have example resulting VERSION strings ?

**Comment by tboerger** [comment]
`REF` is used to detect if it's `HEAD`, `master` or a tag, `SHA` just adds the commit hash. So of course we need both if we want to detect the version automatically.

**Comment by strk** [comment]
And why do we want to distinguish, was my question.
When calling `git describe --always --tag` while in a tag, I do get the tag as the version, which is what we want, while being not in a tag would give you the most recent tag + a count of commits + the actual hash, what else are we missing from that ?

**Comment by tboerger** [comment]
We WANT to have `master=0.0.0|HEAD=0.0.0|tag=1.0.0+commit-sha`, maybe we also want to strip the `v` from the tag name.

**Comment by strk** [comment]
On Mon, Nov 07, 2016 at 04:15:17AM -0800, Thomas Boerger wrote:

> We WANT to have `master=0.0.0|HEAD=0.0.0|tag=1.0.0+commit-sha`, maybe we also want to strip the `v` from the tag name.

I don't get this, why don't you always want an unambiguous
version number, including the "commit-sha" ?

**Comment by tboerger** [comment]
We want a proper version scheme that applies for package managers.

**Comment by strk** [comment]
Maybe I misunderstood your suggestion, as long as the `+commit-sha` is always appended (also for master/HEAD) I'm ok with the proposal

File: main.go [hunk 3]
```diff
@@ -17,7 +17,7 @@ import (
 )
 
 // Version holds the current Gitea version
-const Version = "0.9.99.0915"
+var Version = "0.9.99.0915"
```
Comment (tboerger) [inline main.go:20]: I would set that to `var Version = "0.0.0+master"`

**Review by tboerger — CHANGES_REQUESTED** [review]

**Comment by strk** [comment]
On Mon, Nov 07, 2016 at 07:19:43AM -0800, Thomas Boerger wrote:

>  // Version holds the current Gitea version
> -const Version = "0.9.99.0915"
> +var Version = "0.9.99.0915"
> 
> I would set that to `var Version = "0.0.0+master"`

Based on
https://github.com/go-gitea/gitea/issues/91#issuecomment-258865894
the hard-coded version should probably be "1.0.0+dev"
(target version for current master is 1.0.0)

File: main.go (hunk 3 above)
Comment (tboerger) [reply main.go:20]: If I remember correctly you can not manipulate `const` from outside? Need to confirm that.

Comment (lunny) [reply main.go:20]: A const can not be set via go build flag.

**Comment by tboerger** [comment]
Let's close that in favor of #190

**Pull request closed without merging**